<a href="https://colab.research.google.com/github/JoseAlberto88/Marketing-Analytics-Problem-Assignment/blob/main/Assingment_part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd

file_path = '/content/drive/MyDrive/Marketing analytics/Purchasing_data.xlsx'

# Read with no header assumed, so we can control exactly which rows are what
df_raw = pd.read_excel(file_path, header=None)

# Row 0 = short codes -> use these as the actual column names
short_codes = df_raw.iloc[0]

# Row 1 = long question text -> save as a lookup dictionary (short code -> full question)
question_labels = dict(zip(short_codes, df_raw.iloc[1]))

# Row 2 onward = actual respondent data
df = df_raw.iloc[2:].reset_index(drop=True)
df.columns = short_codes

print(f"Fixed shape: {df.shape}")
print("\nExample lookup:")
print("D15 =", question_labels["D15"])

df.head()

Fixed shape: (4362, 89)

Example lookup:
D15 = Which of the following statements best reflects your current financial lifestyle?


,Q1,End Date,Q01_1,Q01_2,Q01_3,Q01_4,Q01_5,Q01_6,Q01_7,Q01_8,...,D07,D08,D09,D10,D11,D12,D13,D14,D15,D16
0,2021-09-15 09:26:56,2021-09-15 09:32:21,Cash,Check,Credit card,Debit card,Electronic bank transfer,0,0,PayPal,...,1995,25 to 34,Bachelor’s degree,Employed full-time (32 hours or more per week),"Between $25,000 and $49,999","Single, never married",No,"Small city (250,001 to 500,000 inhabitants)",I live paycheck to paycheck and have difficult...,I have no savings.
1,2021-09-15 09:26:07,2021-09-15 09:32:25,Cash,Check,Credit card,Debit card,Electronic bank transfer,0,0,0,...,1968,45 to 54,Associate degree,Self-employed or gig worker,"More than $100,000",Married or domestic partnership,No,"Small town (25,001 to 100,000 inhabitants)",I live paycheck to paycheck but am comfortably...,"Between $2,500 and $5,000"
2,2021-09-15 09:26:05,2021-09-15 09:39:26,Cash,Check,Credit card,Debit card,0,0,Gift/prepaid card,0,...,1956,55 to 64,High school diploma or equivalent (GED),Retired,"Between $75,000 and $99,999",Married or domestic partnership,No,"Small city (250,001 to 500,000 inhabitants)",I live paycheck to paycheck but am comfortably...,"Between $2,500 and $5,000"
3,2021-09-15 09:49:43,2021-09-15 09:54:15,0,0,Credit card,Debit card,0,0,Gift/prepaid card,0,...,1951,65 or older,High school diploma or equivalent (GED),Retired,"Below $25,000",Divorced,No,"Rural or farming area (fewer than 25,000 inhab...",I live paycheck to paycheck but am comfortably...,I have no savings.
4,2021-09-15 09:51:14,2021-09-15 09:56:18,Cash,Check,Credit card,Debit card,Electronic bank transfer,Store cards (cards that can only be used at sp...,Gift/prepaid card,PayPal,...,1952,65 or older,Bachelor’s degree,Retired,"More than $100,000",Married or domestic partnership,"Yes, full-time","Small city (250,001 to 500,000 inhabitants)",I do not live paycheck to paycheck,"More than $25,000"


In [4]:
# ============================================================
# Full column map: short code -> full question text
# ============================================================
import pandas as pd

column_overview = pd.DataFrame({
    "Column Code": df.columns,
    "Question Text": [question_labels.get(col, "N/A") for col in df.columns]
})

pd.set_option('display.max_colwidth', None)
display(column_overview)

,Column Code,Question Text
0,Q1,Start Date
1,End Date,End Date
2,Q01_1,"In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Cash"
3,Q01_2,"In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Check"
4,Q01_3,"In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Credit card"
...,...,...
84,D12,Marital status
85,D13,Are there any children under the age of 18 currently living in your household?
86,D14,Where do you live?
87,D15,Which of the following statements best reflects your current financial lifestyle?


In [5]:
for _, row in column_overview.iterrows():
    print(f"{row['Column Code']}: {row['Question Text']}")

Q1: Start Date
End Date: End Date
Q01_1: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Cash 
Q01_2: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Check 
Q01_3: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Credit card 
Q01_4: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Debit card 
Q01_5: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Electronic bank transfer
Q01_6: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Store cards (cards that can only be used at specific merchants)
Q01_7: In the past 12 months, which of the following methods have you used to make payments? Pl

Let's now inspect teh actual values in the columns we need, using the `df` dataset

In [6]:
# Listing the columns that we are interesting in

digital_cols = ['Q01_8', 'Q01_9', 'Q01_10', 'Q01_11', 'Q01_12', 'Q01_13', 'Q01_14']

for col in digital_cols:
    print(f"--- {col}: {question_labels[col]} ---")
    print(df[col].value_counts(dropna=False))
    print()

print("--- D15:", question_labels['D15'], "---")
print(df['D15'].value_counts(dropna=False))

--- Q01_8: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - PayPal ---
Q01_8
0         2197
PayPal    2119
Paypal      46
Name: count, dtype: int64

--- Q01_9: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Another peer-to-peer payment app (such as Venmo) ---
Q01_9
0                                                   3757
Another peer-to-peer payment app (such as Venmo)     605
Name: count, dtype: int64

--- Q01_10: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - - Apple Pay ---
Q01_10
0            3733
Apple Pay     629
Name: count, dtype: int64

--- Q01_11: In the past 12 months, which of the following methods have you used to make payments? Please select all that apply - Selected Choice - Google Pay ---
Q01_11
0             3747
Google Pay     615
Name: count, dtype: int64

--- Q0

**What the data actually shows**

**Digital payment columns (Q01_8–Q01_14):** Clean pattern,  each is either the literal string `"0"` (not selected) or the option's text label (selected). Easy to binarize. **One data-quality wrinkle:** Q01_8 has inconsistent capitalization , "PayPal" (2,119) vs. "Paypal" (46) are the same answer, just typed differently by Qualtrics/respondents. Doesn't affect our binary flag (both just mean "selected"), but worth noting in your cleaning write-up as an inconsistent-value issue you caught and handled.

**D15 (financial lifestyle):** Three genuine ordinal categories, but a big problem: **2,268 of 4,362 respondents (52%) have no answer at all.** That's more than half your sample. This is a real decision point, not a minor cleanup, since D15 is one of our two clustering variables, this directly determines how much data you actually get to cluster on.

Let's check if this is a random or a structural pattern (such as survey drop-off, skip logic tied to another question) before deciding how to handle it.

In [7]:
# 1. Check if there are extra Qualtrics metadata columns not in the codebook
# (e.g., Progress, Finished, Duration) that might explain drop-off
print(df.columns.tolist())

['Q1', 'End Date', 'Q01_1', 'Q01_2', 'Q01_3', 'Q01_4', 'Q01_5', 'Q01_6', 'Q01_7', 'Q01_8', 'Q01_9', 'Q01_10', 'Q01_11', 'Q01_12', 'Q01_13', 'Q01_14', 'Q01_15', 'Q02', 'Q03', 'Q04', 'Q05', 'Q06', 'Q07_1', 'Q07_2', 'Q07_3', 'Q07_4', 'Q07_5', 'Q07_6', 'Q10', 'Q11_1', 'Q11_2', 'Q11_3', 'Q11_4', 'Q12_1', 'Q12_2', 'Q12_3', 'Q12_4', 'Q12_5', 'Q12_6', 'Q12_7', 'Q12_8', 'Q12_9', 'Q13', 'Q13a', 'Q14', 'C1', 'D01_1', 'D01_4', 'D01_5', 'D01_6', 'D01_7', 'D01_8', 'D01_9', 'D01_10', 'D01_11', 'D01_12', 'D02_1', 'D02_4', 'D02_5', 'D02_6', 'D02_7', 'D02_8', 'D02_9', 'D02_10', 'D02_11', 'D02_12', 'D02_13', 'D02_14', 'D03_1', 'D03_4', 'D03_5', 'D03_6', 'D03_7', 'D03_8', 'D03_9', 'D03_10', 'D04_1', 'D04_2', 'D06', 'D07', 'D08', 'D09', 'D10', 'D11', 'D12', 'D13', 'D14', 'D15', 'D16']


In [8]:
# 2. Does D15 missingness line up with missingness in nearby demographic columns?
# If respondents who skipped D15 also skipped D16, D14, D13, etc., it suggests
# survey abandonment (they stopped before reaching the end), not a D15-specific issue.
nearby_cols = ['C1', 'D11', 'D12', 'D13', 'D14', 'D15', 'D16']

missing_pattern = df[nearby_cols].isna()
print(missing_pattern.sum())
print()

# Cross-tab: of those missing D15, how many are ALSO missing D16 (the very last question)?
d15_missing = df['D15'].isna()
print("Among respondents missing D15:")
print(df.loc[d15_missing, 'D16'].isna().value_counts())
print()
print("Among respondents NOT missing D15:")
print(df.loc[~d15_missing, 'D16'].isna().value_counts())

0
C1     2268
D11    2268
D12    2268
D13    2268
D14    2268
D15    2268
D16    2268
dtype: int64

Among respondents missing D15:
D16
True    2268
Name: count, dtype: int64

Among respondents NOT missing D15:
D16
False    2094
Name: count, dtype: int64


In [9]:
# 3. Check if there's a Progress/Finished/Duration column (common in raw Qualtrics exports)
# that directly tells us who completed the survey vs. dropped off partway
possible_meta_cols = [c for c in df.columns if any(k in str(c).lower() for k in ['progress', 'finish', 'duration', 'status'])]
print("Possible metadata columns found:", possible_meta_cols)

if possible_meta_cols:
    for col in possible_meta_cols:
        print(f"\n--- {col} ---")
        print(df[col].value_counts(dropna=False).head(10))

Possible metadata columns found: []


In [10]:
# 4. Alternative check: does D15 missingness correlate with which Q07/Q12 "version"
# a respondent got? (Since this survey clearly has multiple randomized question-wording
# versions, e.g. Q07_1 through Q07_6, Q12_1 through Q12_9 — a respondent likely only
# answers ONE version, leaving the rest blank by design, not by error.)
q07_cols = ['Q07_1', 'Q07_2', 'Q07_3', 'Q07_4', 'Q07_5', 'Q07_6']
existing_q07 = [c for c in q07_cols if c in df.columns]
print(df[existing_q07].notna().sum())

0
Q07_1    2094
Q07_2    2094
Q07_3    2094
Q07_4    2094
Q07_5    2094
Q07_6    2094
dtype: int64


**What's actually going on**

**All 2,268 "missing D15" respondents are missing C1, D11, D12, D13, D14, D15, and D16**, **every single one of them, together, as one block.** And the reverse holds too: every respondent who has D15 also has all the others. This isn't scattered, random missingness, it's a **hard structural break in the survey.**

Combined with the Q07 finding (exactly 2,094 non-null for every Q07 version column, matching precisely the number of respondents *with* D15), the picture is clear:

- **2,094 respondents completed the entire survey**, including the credit score and demographics block at the end.
- **2,268 respondents dropped off partway through** (somewhere after the credit-card behavior questions, before reaching credit score/demographics), and Qualtrics still exported their partial responses as rows, rather than discarding them.

This is a textbook **survey drop-off / incomplete-response pattern**, not a sensitive-question refusal or a data-quality bug. It's also good news methodologically: it means excluding these rows isn't "losing data we could have kept",  it's **correctly excluding incomplete survey responses**, which is standard, well-justified practice, not an arbitrary judgment call.

***Recommended action**

In [11]:
# Keep only respondents who completed the full survey (reached the demographics block)
df_complete = df[df['D15'].notna()].reset_index(drop=True)

print(f"Original: {df.shape[0]} rows")
print(f"Complete responses only: {df_complete.shape[0]} rows")

Original: 4362 rows
Complete responses only: 2094 rows


This single filter cleanly resolves missingness in **C1, D11, D12, D13, D14, D15, and D16 all at once**,  not just D15, since they're all missing together. Your cleaning write-up can state this plainly: *"2,268 of 4,362 responses (52%) were partial/incomplete, terminating before the credit score and demographics section; these were excluded as incomplete survey responses, leaving 2,094 complete responses for analysis."* That's a clean, defensible sentence for your report.